### PR TITLE
[21797] Add Fast DDS v3 Migration Guide

### PR DESCRIPTION
**Pull request by LuciaEchevarria99**
## Description

Documentation related to PR: https://github.com/eProsima/Fast-DDS/pull/5281

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- N/A: Code snippets related to the added documentation have been provided.
- [x] Documentation tests pass locally.
- N/A: Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] CI passes without warnings or errors.

### CODE REVIEW

File: docs/notes/migration_guide.rst [hunk 1]
```diff
@@ -0,0 +1,491 @@
+.. include:: ../03-exports/aliases.include
+.. include:: ../03-exports/aliases-api.include
+.. include:: ../03-exports/roles.include
+
+Migration Guide from Fast DDS v2 to Fast DDS v3
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:5]: Title too long for this section, I would keep it shorter
```suggestion
Migration Guide to Fast DDS v3
```

File: docs/notes/migration_guide.rst [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+===============================================
+
+This document aims to help during the migration process from eProsima *Fast DDS version* 2 to *Fast DDS version* 3.
+For more information about all the updates, please refer to the [release notes](https://fast-dds.docs.eprosima.com/en/latest/notes/notes.html).
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:9]: I would add a reference in supported_versions.rst as `versions_v3` to tag it here
```suggestion
For more information about all the updates, please refer to the :ref:`release notes <release_notes_v3>`.
```

File: docs/notes/migration_guide.rst [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+
+Step 1: Update the Package Name and CMake Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. CMake Project Name: Rename the CMake project from ``fastrtps`` to ``fastdds``.
+2. Environment Variables:
+
+   * Rename ``FASTRTPS_DEFAULT_PROFILES_FILE`` to ``FASTDDS_DEFAULT_PROFILES_FILE``.
+   * The configuration file for loading profiles is now ``DEFAULT_FASTDDS_PROFILES.xml``.
+3. Update CMake File Names on Windows:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:36]:
```suggestion
3. Update CMakeFile names on Windows:
```

File: docs/notes/migration_guide.rst [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+- :ref:`step-7-update-api-methods`
+- :ref:`step-8-update-structs-enums-and-variables`
+- :ref:`step-9-refactor-examples`
+
+.. _step-1-update-the-package-name-and-cmake-configuration:
+
+Step 1: Update the Package Name and CMake Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. CMake Project Name: Rename the CMake project from ``fastrtps`` to ``fastdds``.
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:31]:
```suggestion
1. CMake project name: Rename the CMake project from ``fastrtps`` to ``fastdds``.
```

File: docs/notes/migration_guide.rst [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+2. Environment Variables:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:32]:
```suggestion
2. Environment variables:
```

File: docs/notes/migration_guide.rst [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+Refer to the :ref:`library deprendencies table <dependencies_compatibilities_build_system_dependencies>` to verify version compatibility for all Fast DDS library dependencies.
+
+.. _step-3-ensure-compability-with-related-products:
+
+Step 3: Ensure Compability with Related Products
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Fast DDS v3 requires Fast DDS Gen v4 for code generation. Make sure to regenerate types using this compatible version.
+
+For other compatibility requirements with related products (e.g., Shapes Demo, Discovery Server), refer to the table of [products compatibility](https://fast-dds.docs.eprosima.com/en/latest/notes/versions.html#eprosima-products-compatibility), which outlines version compatibility across the Fast DDS ecosystem.
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:58]:
```suggestion
For other compatibility requirements with related products (e.g., Shapes Demo, Discovery Server), refer to the table of :ref:`products compatibility <dependencies_compatibilities_product_compatibility>`, which outlines version compatibility across the Fast DDS ecosystem.
```

File: docs/notes/migration_guide.rst [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+- :ref:`step-4-apply-namespace-changes`
+- :ref:`step-5-migrate-public-headers`
+- :ref:`step-6-handle-removed-or-private-headers`
+- :ref:`step-7-update-api-methods`
+- :ref:`step-8-update-structs-enums-and-variables`
+- :ref:`step-9-refactor-examples`
+
+.. _step-1-update-the-package-name-and-cmake-configuration:
+
+Step 1: Update the Package Name and CMake Configuration
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:28]:
```suggestion
Step 1: Update the package name and CMake configuration
```

File: docs/notes/migration_guide.rst [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+   * The configuration file for loading profiles is now ``DEFAULT_FASTDDS_PROFILES.xml``.
+3. Update CMake File Names on Windows:
+
+   * Rename ``fastrtps.manifest.in`` to ``fastdds.manifest.in``.
+   * Rename ``fastrtps-config.cmake`` to ``fastdds-config.cmake``.
+   * Rename ``fastrtps.rc`` to ``fastdds.rc``.
+
+.. _step-2-update-dependencies:
+
+Step 2: Update Dependencies
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:44]:
```suggestion
Step 2: Update dependencies
```

File: docs/notes/migration_guide.rst [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Fast DDS v3 is only compatible with Fast CDR v2.
+If you are not using Fast CDR as :ref:`third-party <third-party_libraries-options>`, please ensure that your local dependencies are up-to-date.
+Refer to the :ref:`library deprendencies table <dependencies_compatibilities_build_system_dependencies>` to verify version compatibility for all Fast DDS library dependencies.
+
+.. _step-3-ensure-compability-with-related-products:
+
+Step 3: Ensure Compability with Related Products
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:53]:
```suggestion
Step 3: Ensure compatibility with related products
```

File: docs/notes/migration_guide.rst [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Fast DDS v3 requires Fast DDS Gen v4 for code generation. Make sure to regenerate types using this compatible version.
+
+For other compatibility requirements with related products (e.g., Shapes Demo, Discovery Server), refer to the table of [products compatibility](https://fast-dds.docs.eprosima.com/en/latest/notes/versions.html#eprosima-products-compatibility), which outlines version compatibility across the Fast DDS ecosystem.
+Verify these versions and update accordingly to avoid any integration issues.
+
+.. _step-4-apply-namespace-changes:
+
+Step 4: Apply Namespace Changes
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:63]:
```suggestion
Step 4: Apply namespace changes
```

File: docs/notes/migration_guide.rst [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. Namespace Migration:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:66]:
```suggestion
1. Namespace migration:
```

File: docs/notes/migration_guide.rst [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+
+   * Update all ``eprosima::fastrtps::`` namespace references to ``eprosima::fastdds::``.
+   * Move built-in topics ``SubscriptionBuiltinTopicData``, ``PublicationBuiltinTopicData``, and ``ParticipantBuiltinTopicData`` from ``fastdds::dds::builtin::`` to ``fastdds::dds::``.
+   * Move ``Duration_t`` and ``c_TimeInfinite`` references to ``dds::``.
+   * Move ``Time_t.hpp`` references from ``eprosima::fastdds::`` to ``eprosima::fastdds::dds``.
+
+   Ensure you update these namespace references across your code to avoid compilation errors.
+
+2. Renamed Types:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:75]:
```suggestion
2. Renamed types:
```

File: docs/notes/migration_guide.rst [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+
+   * Change ``EventKindBits::`` references to ``EventKind::``.
+   * Change ``EventKindEntityId::`` references to ``EntityId::``.
+   * Change ``StatisticsEventKind::`` references to ``statistics::EventKind::``.
+
+   Refactor the type references as outlined above to maintain compatibility with the new version.
+
+.. _step-5-migrate-public-headers:
+
+Step 5: Migrate Public Headers
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:85]:
```suggestion
Step 5: Migrate public headers
```

File: docs/notes/migration_guide.rst [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. File Location headers:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:88]:
```suggestion
1. Header files location:
```

File: docs/notes/migration_guide.rst [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+
+   All the headers in ``include/fastrtps`` were migrated to ``include/fastdds``.
+   In particular, the following list includes headers that have been relocated to different paths or whose implementations have been incorporated into other headers:
+
+   .. list-table::
+      :header-rows: 1
+
+      * - Fast DDS 2.x file *include* path
+        - Fast DDS v3.0.0 file *include* path
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:97]:
```suggestion
      * - Fast DDS v2 file *include* path
        - Fast DDS v3 file *include* path
```

File: docs/notes/migration_guide.rst [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+      * - ``fastrtps/transport/UDPTransportDescritpor.h``
+        - ``fastdds/rtps/transport/UDPTransportDescritpor.hpp``
+      * - ``fastrtps/transport/TCPTransportDescritpor.h``
+        - ``fastdds/rtps/transport/TCPTransportDescritpor.hpp``
+      * - ``fastdds/rtps/common/Time_t.hpp in namespace{fastdds}``
+        - ``fastdds/dds/core/Time_t.hpp in namespace{fastdds::dds}``
+
+   Also, the ``fixed_size_string.hpp`` implementation has been migrated from Fast DDS package to Fast CDR.
+
+2. File Extensions:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:141]:
```suggestion
2. File extensions:
```

File: docs/notes/migration_guide.rst [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+
+   Rename file extensions from `.h` to `.hpp`.
+
+.. _step-6-handle-removed-or-private-headers:
+
+Step 6: Handle Removed or Private Headers
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:147]:
```suggestion
Step 6: Handle removed or private headers
```

File: docs/notes/migration_guide.rst [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+* ServerAttributes.hpp
+* TopicAttributes.hpp
+* TypeLookupService.hpp
+
+If your project previously included any of these headers, you will need to modify your implementation.
+Since these headers are now private, you should replace their usage with public alternatives or refactor the related code to ensure it does not depend on private headers.
+
+.. _step-7-update-api-methods:
+
+Step 7: Update API Methods
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:203]:
```suggestion
Step 7: Update API methods
```

File: docs/notes/migration_guide.rst [hunk 19]
```diff
@@ -0,0 +1,491 @@
...
+   * - ``DynamicPubSubType::getSerializedSizeProvider(const void* const, DataRepresentationId_t)``
+     - ``DynamicPubSubType::calculate_serialized_size(const void* const, DataRepresentationId_t)``
+
+Review your code for any APIs marked with the ``FASTDDS_DEPRECATED`` and ``FASTDDS_TODO_BEFORE`` macros.
+Note that these deprecated APIs have been removed in Fast DDS v3.
+Make the necessary updates to your implementation to ensure compatibility with the new version.
+
+.. _step-8-update-structs-enums-and-variables:
+
+Step 8: Update Structs, Enums, and Variables
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:347]:
```suggestion
Step 8: Update structs, enums, and variables
```

File: docs/notes/migration_guide.rst [hunk 20]
```diff
@@ -0,0 +1,491 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As part of the Fast DDS migration, several structs, enums, and variables have been updated. You will need to modify your code to reflect these changes:
+
+1. Enum and Variable Changes:
+
+    * ``DiscoveryProtocol_t`` is now ``DiscoveryProtocol``.
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:354]:
```suggestion
    * Rename ``DiscoveryProtocol_t`` to ``DiscoveryProtocol``.
```

File: docs/notes/migration_guide.rst [hunk 21]
```diff
@@ -0,0 +1,491 @@
...
+    * ``ParticipantBuiltinTopicData`` has been extended to include the product version and fields from ``ParticipantProxyData``.
+
+3. Other Struct Changes:
+
+    * ``SendBuffersAllocationAttributes`` has a new attribute to define the allocation configuration of the ``NetworkBuffers``.
+    * ``TypeConsistencyQos`` has been removed from ``DataReader``, and the ``TypeConsistencyEnforcementQosPolicy`` and ``DataRepresentationQosPolicy`` have been added.
+
+.. _step-9-refactor-examples:
+
+Step 9: Refactor Examples
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:379]:
```suggestion
Step 9: Examples refactor
```

File: docs/notes/migration_guide.rst [hunk 22]
```diff
@@ -0,0 +1,491 @@
...
+All examples have been refactored to follow a consistent structure across the Fast DDS project.
+This includes renaming files, restructuring classes, and updating the overall format.
+Additionally, it is important to note that some examples have been removed, renamed, or had significant changes to their options and configurations.
+If you have integrated any of these examples into your own implementation, carefully review the updated examples to ensure compatibility with your project.
+
+All the examples have been refactored to follow the same structure:
+
+* File names, guards, and classes follow new format.
+* Detailed and well-formed README.md with example explanation.
+* Example structured in applications, stopped by `SIGTERM` signal.
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:391]:
```suggestion
* Example structured in applications, stopped by ``SIGTERM`` signal.
```

File: docs/notes/migration_guide.rst [hunk 23]
```diff
@@ -0,0 +1,491 @@
...
+The table below contains the list of API changes, showing the previous methods and the corresponding new ones introduced in Fast DDS v3.
+The new API methods achieve the same functionality, even though the signature of the method is different from the deprecated one.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Deprecated methods
+     - New methods
+   * - ``xmlparser::XMLProfileManager::library_settings(LibrarySettingsAttributes&)``
+     - ``DomainParticipantFactory::get_instance()->set_library_settings(const LibrarySettings&)``
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:215]: Ideally, we should reference the API reference in the new methods, but it is a lot of work...

File: docs/notes/migration_guide.rst [hunk 24]
```diff
@@ -0,0 +1,491 @@
...
+If you have integrated any of these examples into your own implementation, carefully review the updated examples to ensure compatibility with your project.
+
+All the examples have been refactored to follow the same structure:
+
+* File names, guards, and classes follow new format.
+* Detailed and well-formed README.md with example explanation.
+* Example structured in applications, stopped by `SIGTERM` signal.
+
+Hello World
+"""""""""""
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:394]: What do you think about linking the examples here? Apply in the remain examples if considered

```suggestion
`Hello World <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/hello_world>`_
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```

**Review by JesusPoderoso — CHANGES_REQUESTED** [review]

File: docs/notes/migration_guide.rst (hunk 23 above)
Comment (LuciaEchevarria99) [reply docs/notes/migration_guide.rst:215]: But I wouldn't be able to reference the deleted methods, right?

Comment (JesusPoderoso) [reply docs/notes/migration_guide.rst:215]: Exactly

File: docs/notes/migration_guide.rst [hunk 25]
```diff
@@ -0,0 +1,510 @@
...
+* Loans mechanism for data management.
+* Option to select all delivery mechanisms.
+
+`Discovery Server <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/discovery_server>`_
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+Refactor the DiscoveryServerExample example with the current new example format.
+
+`Flow Controller <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/flow_controller>`_
+""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:475]:
```suggestion
`Flow Controller <https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/flow_control>`_
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```

File: docs/notes/migration_guide.rst (hunk 10 above)
Comment (JesusPoderoso) [reply docs/notes/migration_guide.rst]: Not applied

File: docs/notes/migration_guide.rst [hunk 26]
```diff
@@ -0,0 +1,510 @@
...
+Step 1: Update the package name and CMake configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+1. CMake project name: Rename the CMake project from ``fastrtps`` to ``fastdds``.
+2. Environment variables:
+
+   * Rename ``FASTRTPS_DEFAULT_PROFILES_FILE`` to ``FASTDDS_DEFAULT_PROFILES_FILE``.
+   * The configuration file for loading profiles is now ``DEFAULT_FASTDDS_PROFILES.xml``.
+
+3. Update CMakeFile names on Windows:
```
Comment (JesusPoderoso) [inline docs/notes/migration_guide.rst:37]: I messed up previous revision here, sorry
```suggestion
3. Update CMake file names on Windows:
```

**Review by JesusPoderoso — CHANGES_REQUESTED** [review]

File: docs/notes/migration_guide.rst (hunk 23 above)
Comment (rsanchez15) [reply docs/notes/migration_guide.rst:226]: This will be done in a follow-up PR. See [21797] (https://eprosima.easyredmine.com/issues/21797)

**Pull request merged**